nilearn/regions/region_extractor.py

```diff
@@ -336,3 +412,135 @@ def fit(self, X=None, y=None):
+    labels_img = check_niimg_3d(labels_img)
+    labels_data = _safe_get_data(labels_img, ensure_finite=True)
+    affine = labels_img.get_affine()
+
+    check_unique_labels = np.unique(labels_data)
+
+    if min_size is not None and not isinstance(min_size, numbers.Number):
+        raise ValueError("Expected 'min_size' to be specified as integer. "
+                         "You provided {0}".format(min_size))
+    if not isinstance(connect_diag, bool):
+        raise ValueError("'connect_diag' must be specified as True or False. "
+                         "You provided {0}".format(connect_diag))
+    if np.any(check_unique_labels < 0):
+        raise ValueError("The 'labels_img' you provided has unknown/negative "
+                         "integers as labels {0} assigned to regions. "
+                         "All regions in an image should have positive "
+                         "integers assigned as labels."
+                         .format(check_unique_labels))
+
+    unique_labels = set(check_unique_labels)
+    # check for background label indicated as 0
+    if np.any(check_unique_labels == 0):
+        unique_labels.remove(0)
+
+    if labels is not None:
+        if (not isinstance(labels, collections.Iterable) or
+                isinstance(labels, _basestring)):
+            labels = [labels, ]
+        if len(unique_labels) != len(labels):
+            raise ValueError("The number of labels: {0} provided as input "
+                             "in labels={1} does not match with the number "
+                             "of unique labels in labels_img: {2}. "
+                             "Please provide appropriate match with unique "
+                             "number of labels in labels_img."
+                             .format(len(labels), labels, len(unique_labels)))
+        new_names = []
+
+    if labels is None:
+        this_labels = [None] * len(unique_labels)
+    else:
+        this_labels = labels
+
+    new_labels_data = np.zeros(labels_data.shape, dtype=np.int)
+    current_max_label = 0
+    for label_id, name in zip(unique_labels, this_labels):
+        this_label_mask = (labels_data == label_id)
+        # Extract regions assigned to each label id
+        if connect_diag:
+            structure = np.ones((3, 3, 3), dtype=np.int)
+            regions, this_n_labels = ndimage.label(
+                this_label_mask.astype(np.int), structure=structure)
+        else:
+            regions, this_n_labels = ndimage.label(this_label_mask.astype(np.int))
+
+        if min_size is not None:
+            index = np.arange(this_n_labels + 1)
+            this_label_mask = this_label_mask.astype(np.int)
+            regions = _remove_small_regions(regions, this_label_mask,
+                                            index, affine, min_size=min_size)
+            this_n_labels = regions.max()
+
+        cur_regions = regions[regions != 0] + current_max_label
+        new_labels_data[regions != 0] = cur_regions
```

Comment:
@AlexandreAbraham Doesn't look like the labels are as expected. Are we missing something here ?
You meant replace with two lines ?